### PR TITLE
crimson/osd/pg: reset the snap mapper's backend when pg interval changes

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Fixes: https://tracker.ceph.com/issues/65679
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by Matan-B** [comment]
```
DEBUG active+clean+snaptrim  ObjectContextLoader::get_or_load_obc: cache hit on 3:6bb7f7ea:::scephqa03.cpp.bjat.qianxin-inc.cn204052-10:head
/src/crimson/osd/object_context_loader.cc:108:  Assertion `cit != std::end(ss.clones)' failed.
```
Can you please explain how this assert relates to the stale cache?

File: src/common/map_cacher.hpp [hunk 1]
```diff
@@ -85,6 +85,12 @@ class MapCacher {
 public:
   MapCacher(StoreDriver<K, V> *driver) : driver(driver) {}
 
+#ifdef WITH_SEASTAR
```
Comment (athanatos) [inline src/common/map_cacher.hpp:88]: This behavior is not specific to seastar, remove the ifdef.

File: src/common/sharedptr_registry.hpp [hunk 2]
```diff
@@ -61,6 +61,13 @@ class SharedPtrRegistry {
     waiting(0)
   {}
 
+#ifdef WITH_SEASTAR
```
Comment (athanatos) [inline src/common/sharedptr_registry.hpp:64]: This behavior is not specific to seastar, remove the ifdef.

File: src/osd/SnapMapper.h [hunk 3]
```diff
@@ -356,6 +356,13 @@ class SnapMapper : public Scrub::SnapMapReaderI {
     return prefix_itr;
   }
 
+#ifdef WITH_SEASTAR
```
Comment (athanatos) [inline src/osd/SnapMapper.h:359]: This behavior is not specific to seastar, remove the ifdef.

**Review by athanatos — APPROVED** [review]
LGTM other than removing the ifdef's

**Comment by xxhdx1985126** [comment]
> ```
> DEBUG active+clean+snaptrim  ObjectContextLoader::get_or_load_obc: cache hit on 3:6bb7f7ea:::scephqa03.cpp.bjat.qianxin-inc.cn204052-10:head
> /src/crimson/osd/object_context_loader.cc:108:  Assertion `cit != std::end(ss.clones)' failed.
> ```
> 
> Can you please explain how this assert relates to the stale cache?

This is a little bit tricky.

1. Snaps `snap1` and `snap2` on object `obj1` maps to the same clone `c`
2. Snap `snap1` is being trimmed on PG `P`;
3. Before `snap1` of `obj1` is trimmed and after the `obj1 <-> snap1` mapping is removed from the snap mapper, `P`'s interval changed;
4. Snap `snap2` is trimmed on `obj1` and caused the `c` to be removed
5. The osd got restarted and find that `snap1` of `obj1` hasn't been trimmed
6. The osd tries to trim `snap1` on `obj1` and meet the assertion `cit != std::end(ss.clones)`

**Review by Matan-B — APPROVED** [review]
> This is a little bit tricky.
> 
> 1. Snaps `snap1` and `snap2` on object `obj1` maps to the same clone `c`
> 2. Snap `snap1` is being trimmed on PG `P`;
> 3. Before `snap1` of `obj1` is trimmed and after the `obj1 <-> snap1` mapping is removed from the snap mapper, `P`'s interval changed;
> 4. Snap `snap2` is trimmed on `obj1` and caused the `c` to be removed
> 5. The osd gets restarted and find that `snap1` of `obj1` hasn't been trimmed
> 6. The osd tries to trim `snap1` on `obj1` and meet the assertion `cit != std::end(ss.clones)`

Ah, thanks! LGTM besides comments left by @athanatos

**Comment by xxhdx1985126** [comment]
All done:-)

**Comment by Matan-B** [comment]
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-57125

**Comment by Matan-B** [comment]
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-june-9/

**Comment by Matan-B** [comment]
jenkins test windows

**Comment by Matan-B** [comment]
jenkins test windows

**Comment by Matan-B** [comment]
Merging based on (21/21 passed):

https://pulpito.ceph.com/matan-2024-06-09_09:28:06-crimson-rados-wip-matanb-crimson-testing-june-9-distro-crimson-smithi/
rerun dead jobs:
https://pulpito.ceph.com/matan-2024-06-09_10:57:06-crimson-rados-wip-matanb-crimson-testing-june-9-distro-crimson-smithi/

**Comment by Matan-B** [comment]
jenkins test windows

**Comment by xxhdx1985126** [comment]
Rebased.

**Comment by Matan-B** [comment]
jenkins test api

**Comment by Matan-B** [comment]
jenkins test windows

**Comment by Matan-B** [comment]
jenkins test api

**Comment by Matan-B** [comment]
jenkins test windows

**Comment by Matan-B** [comment]
jenkins test windows

**Comment by Matan-B** [comment]
jenkins test windows

**Comment by Matan-B** [comment]
jenkins test windows

**Comment by Matan-B** [comment]
jenkins test windows

**Comment by Matan-B** [comment]
jenkins test windows

**Comment by Matan-B** [comment]
jenkins test windows

**Comment by Matan-B** [comment]
jenkins test windows

**Comment by Matan-B** [comment]
jenkins test windows

**Comment by Matan-B** [comment]
rebased on latest main

**Comment by Matan-B** [comment]
jenkins test windows

**Comment by Matan-B** [comment]
windows test is not playing along

**Comment by Matan-B** [comment]
jenkins test windows

**Comment by Matan-B** [comment]
jenkins test windows

**Comment by Matan-B** [comment]
```
src/test/common/CMakeFiles/unittest_sharedptr_registry.dir/test_sharedptr_registry.cc.obj -c /home/ubuntu/ceph/src/test/common/test_sharedptr_registry.cc
In file included from /home/ubuntu/ceph/src/test/common/test_sharedptr_registry.cc:26:
/home/ubuntu/ceph/src/common/sharedptr_registry.hpp:65:5: error: use of undeclared identifier 'ceph_assert'
    ceph_assert(!waiting);
    ^
1 error generated.
```

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by Matan-B** [comment]
@rzarzynski, common change here though Crimson is the only user meaning no actual impact to classic. Can this one be merged?

**Review by rzarzynski — APPROVED** [review]

**Pull request merged**